File: api/schemas/accounts.py
"""Pydantic schemas for the accounts router."""
from datetime import date

from pydantic import BaseModel


class AccountTypeResponse(BaseModel):
    id: int
    name: str
    is_pension: bool
    in_use: bool


class AccountEntryRequest(BaseModel):
    account_type_id: int
    entry_date: date
    balance: float
    currency: str = "GBP"
    exchange_rate: float = 1.0


class AccountEntryResponse(BaseModel):
    id: int
    account_type_id: int
    entry_date: date
    balance: float
    currency: str


class EntryItemResponse(BaseModel):
    entry_id: int
    type_id: int
    type_name: str
    balance: float


class HistoryDayResponse(BaseModel):
    date: str
    total: float
    entries: list[EntryItemResponse]

File: client/src/lib/api/accounts.ts
import { apiClient } from '@/lib/apiClient';

export interface AccountTypeResponse {
  id: number;
  name: string;
  is_pension: boolean;
  in_use: boolean;
}

export interface AccountEntryRequest {
  account_type_id: number;
  entry_date: string;    // ISO 'yyyy-MM-dd' — Date serialised by BalanceEntryForm
  balance: number;
  currency?: string;     // defaults to 'GBP' on server
  exchange_rate?: number;
}

export interface EntryItemResponse {
  entry_id: number;
  type_id: number;
  type_name: string;
  balance: number;
}

export interface HistoryDayResponse {
  date: string;
  total: number;
  entries: EntryItemResponse[];
}

export const accountsApi = {
  listTypes: (): Promise<AccountTypeResponse[]> =>
    apiClient.get('/api/accounts/types').then(r => r.data),

  createType: (name: string): Promise<AccountTypeResponse> =>
    apiClient.post('/api/accounts/types', { name }).then(r => r.data),

  updateType: (id: number, name: string): Promise<AccountTypeResponse> =>
    apiClient.put(`/api/accounts/types/${id}`, { name }).then(r => r.data),

  deleteType: (id: number): Promise<void> =>
    apiClient.delete(`/api/accounts/types/${id}`).then(r => r.data),

  createEntry: (body: AccountEntryRequest): Promise<unknown> =>
    apiClient.post('/api/accounts/entries', body).then(r => r.data),

  getHistory: (): Promise<HistoryDayResponse[]> =>
    apiClient.get('/api/accounts/history').then(r => r.data),
};
